src/server/game/Behavior/ActionScript/ActionThreadStep.h

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (C) 2008-2018 TrinityCore <https://www.trinitycore.org/>
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation; either version 2 of the License, or (at your
+ * option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef TRINITY_ACTIONTHREADSTEP_H
+#define TRINITY_ACTIONTHREADSTEP_H
+
+#include "ActionScriptDefines.h"
+#include "Optional.h"
+#include <memory>
+
+class ActionThread;
+struct ActionScriptStep;
+class TC_GAME_API ActionThreadStep
+{
+    public:
+        virtual ~ActionThreadStep() {}
+
+        static std::unique_ptr<ActionThreadStep> StepTo(ActionThread const& thread, ActionScriptStep const& stepTemplate);
+
+        virtual void Initialize() = 0;
+        virtual Optional<bool> Evaluate() = 0;
+        virtual void Abort() = 0;
+
+        ActionThread const& GetThread() const { return _thread; }
+        ActionScriptStep const& GetTemplate() const { return _template; }
+
+    protected:
+        ActionThreadStep(ActionThread const& thread, ActionScriptStep const& stepTemplate) : _thread(thread), _template(stepTemplate) {}
+
+        ActionThread const& _thread;
+        ActionScriptStep const& _template;
```

Comment:
Historically, not having access to (something like) the `ActionThread` has always turned out to be a maintenance nightmare. A single extra pointer is not going to kill us.

`ActionScriptStep` is only stored here, not in the thread. We need access to it.

Reply:
That's 2 pointers already which are going to be duplicated on every creature in the world using ActionScript. Well, I guess as long as only the "current" steps are being stored, the memory usage might be within reason, but if at any point it's going to be like SAI, where the entire script is copied into every of its instances because it might need to be changed (SMART_ACTION_CREATE_TIMED_EVENT, SMART_ACTION_INSTALL_AI_TEMPLATE etc), this could blow up.

I still think it's inefficient to store the reference to thread if the step can never be interacted with by anything other than the thread. Store just the state of the step, pass everything else needed for it to function as parameter(s). Unless you intend to do something like `unit->GetCurrentStep().Cancel()`, which would require the step to have all the necessary information (thread, namely) to function in such independent context, but IMO that's not the best design.

But sure, this kind of optimization can be done when the design is finalized and proven to be working, if it doesn't clash any of the implemented functionality.

Reply:
By storing refs you also just implicitly deleted the other special member functions, maybe not too handy for later.